clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp

```diff
@@ -1289,6 +1289,38 @@ static mlir::Value buildPointerArithmetic(CIRGenFunction &CGF,
 }
 
 mlir::Value ScalarExprEmitter::buildMul(const BinOpInfo &Ops) {
+  if (Ops.CompType->isSignedIntegerOrEnumerationType()) {
+    switch (CGF.getLangOpts().getSignedOverflowBehavior()) {
+    case LangOptions::SOB_Defined:
+      if (!CGF.SanOpts.has(SanitizerKind::SignedIntegerOverflow))
+        return Builder.createMul(Ops.LHS, Ops.RHS);
+      [[fallthrough]];
+    case LangOptions::SOB_Undefined:
+      if (!CGF.SanOpts.has(SanitizerKind::SignedIntegerOverflow))
+        return Builder.createNSWMul(Ops.LHS, Ops.RHS);
+      [[fallthrough]];
+    case LangOptions::SOB_Trapping:
+      if (CanElideOverflowCheck(CGF.getContext(), Ops))
+        return Builder.createNSWMul(Ops.LHS, Ops.RHS);
+      llvm_unreachable("NYI");
+    }
+  }
+  if (Ops.FullType->isConstantMatrixType()) {
+    llvm_unreachable("NYI");
+  }
+  if (Ops.CompType->isUnsignedIntegerType() &&
+      CGF.SanOpts.has(SanitizerKind::UnsignedIntegerOverflow) &&
+      !CanElideOverflowCheck(CGF.getContext(), Ops))
+    llvm_unreachable("NYI");
+
+  if (Ops.CompType->isFloatingType()) {
```

Comment:
You also need to check for vectors here, maybe be worth adding a `isFPOrFPVectorTy` like function (see how `isAnyFloatingPointType` is defined).